File: trunk/incubators/net.sf.j2s.java.org.eclipse.osgi/src/org/eclipse/core/runtime/internal/stats/ResourceBundleStats.java
/*******************************************************************************
 * Copyright (c) 2000, 2004 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 * 
 * Contributors:
 *     IBM Corporation - initial API and implementation
 *******************************************************************************/
package org.eclipse.core.runtime.internal.stats;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.*;

/**
 * BundleStats is used to represent information about loaded bundle. A
 * bundlestats instance represents only one bundle.
 */

public class ResourceBundleStats {
	private String pluginId; // the plugin loading this bundle
	private String fileName; // the filename of the bundle 
	private int keyCount = 0; // number of keys in the bundle
	private int keySize = 0; // size of the keys in the bundle
	private int valueSize = 0; // size of the values in the bundle 
	private long hashSize = 0; // size of the hashtable
	private long fileSize = 0;

	private static int sizeOf(String value) {
		return 44 + (2 * value.length());
	}

	private static int sizeOf(Properties value) {
		return (int) Math.round(44 + (16 + (value.size() * 1.25 * 4)) + (24 * value.size()));
	}

	public ResourceBundleStats(String pluginId, String fileName, URL input) {
		this.pluginId = pluginId;
		this.fileName = fileName;
		initialize(input);
	}

	public ResourceBundleStats(String pluginId, String fileName, ResourceBundle bundle) {
		this.pluginId = pluginId;
		this.fileName = fileName;
		initialize(bundle);
	}

	/**
	 * Compute the size of bundle
	 */
	private void initialize(ResourceBundle bundle) {
		for (Enumeration keys = bundle.getKeys(); keys.hasMoreElements();) {
			String key = (String) keys.nextElement();
			keySize += sizeOf(key);
			valueSize += sizeOf(bundle.getString(key));
			keyCount++;
		}
	}

	/**
	 * Compute the size of stream which represents a property file
	 */
	private void initialize(URL url) {
		InputStream stream = null;
		Properties props = new Properties();
		try {
			try {
				stream = url.openStream();
				fileSize = stream.available();
				props.load(stream);
				for (Iterator iter = props.keySet().iterator(); iter.hasNext();) {
					String key = (String) iter.next();
					keySize += sizeOf(key);
					valueSize += sizeOf(props.getProperty(key));
					keyCount++;
				}
				hashSize = sizeOf(props);
			} finally {
				if (stream != null)
					stream.close();
			}
		} catch (IOException e) {
			// ignore exceptions as they will be handled when the stream 
			// is loaded for real.   See callers.
		}
	}

	public long getHashSize() {
		return hashSize;
	}

	public int getKeyCount() {
		return keyCount;
	}

	public String getPluginId() {
		return pluginId;
	}

	public int getKeySize() {
		return keySize;
	}

	public int getValueSize() {
		return valueSize;
	}

	public long getTotalSize() {
		return keySize + valueSize + hashSize;
	}

	public String getFileName() {
		return fileName;
	}

	public long getFileSize() {
		return fileSize;
	}
}

File: trunk/incubator/net.sf.j2s.test.mauve/j2slib/org/eclipse/core/runtime/internal/stats/ResourceBundleStats.js
Clazz.declarePackage ("org.eclipse.core.runtime.internal.stats");
Clazz.load (null, "org.eclipse.core.runtime.internal.stats.ResourceBundleStats", ["java.util.Properties"], function () {
c$ = Clazz.decorateAsClass (function () {
this.pluginId = null;
this.fileName = null;
this.keyCount = 0;
this.keySize = 0;
this.valueSize = 0;
this.hashSize = 0;
this.fileSize = 0;
Clazz.instantialize (this, arguments);
}, org.eclipse.core.runtime.internal.stats, "ResourceBundleStats");
c$.sizeOf = Clazz.defineMethod (c$, "sizeOf", 
($fz = function (value) {
return 44 + (2 * value.length);
}, $fz.isPrivate = true, $fz), "~S");
c$.sizeOf = Clazz.defineMethod (c$, "sizeOf", 
($fz = function (value) {
return Math.round (44 + (16 + (value.size () * 1.25 * 4)) + (24 * value.size ()));
}, $fz.isPrivate = true, $fz), "java.util.Properties");
Clazz.makeConstructor (c$, 
function (pluginId, fileName, input) {
this.pluginId = pluginId;
this.fileName = fileName;
this.initialize (input);
}, "~S,~S,java.net.URL");
Clazz.makeConstructor (c$, 
function (pluginId, fileName, bundle) {
this.pluginId = pluginId;
this.fileName = fileName;
this.initialize (bundle);
}, "~S,~S,java.util.ResourceBundle");
Clazz.defineMethod (c$, "initialize", 
($fz = function (bundle) {
for (var keys = bundle.getKeys (); keys.hasMoreElements (); ) {
var key = keys.nextElement ();
this.keySize += org.eclipse.core.runtime.internal.stats.ResourceBundleStats.sizeOf (key);
this.valueSize += org.eclipse.core.runtime.internal.stats.ResourceBundleStats.sizeOf (bundle.getString (key));
this.keyCount++;
}
}, $fz.isPrivate = true, $fz), "java.util.ResourceBundle");
Clazz.defineMethod (c$, "initialize", 
($fz = function (url) {
var stream = null;
var props =  new java.util.Properties ();
try {
try {
stream = url.openStream ();
this.fileSize = stream.available ();
props.load (stream);
for (var iter = props.keySet ().iterator (); iter.hasNext (); ) {
var key = iter.next ();
this.keySize += org.eclipse.core.runtime.internal.stats.ResourceBundleStats.sizeOf (key);
this.valueSize += org.eclipse.core.runtime.internal.stats.ResourceBundleStats.sizeOf (props.getProperty (key));
this.keyCount++;
}
this.hashSize = org.eclipse.core.runtime.internal.stats.ResourceBundleStats.sizeOf (props);
} finally {
if (stream != null) stream.close ();
}
} catch (e) {
if (Clazz.instanceOf (e, java.io.IOException)) {
} else {
throw e;
}
}
}, $fz.isPrivate = true, $fz), "java.net.URL");
Clazz.defineMethod (c$, "getHashSize", 
function () {
return this.hashSize;
});
Clazz.defineMethod (c$, "getKeyCount", 
function () {
return this.keyCount;
});
Clazz.defineMethod (c$, "getPluginId", 
function () {
return this.pluginId;
});
Clazz.defineMethod (c$, "getKeySize", 
function () {
return this.keySize;
});
Clazz.defineMethod (c$, "getValueSize", 
function () {
return this.valueSize;
});
Clazz.defineMethod (c$, "getTotalSize", 
function () {
return this.keySize + this.valueSize + this.hashSize;
});
Clazz.defineMethod (c$, "getFileName", 
function () {
return this.fileName;
});
Clazz.defineMethod (c$, "getFileSize", 
function () {
return this.fileSize;
});
});
